# Analiza pliku fastq

# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install("ShortRead")

library(R.utils)
library(ShortRead)
library(stringr)
library(NCmisc)

proj.dir = "~/Desktop/HT projects/Rosalind/bin"
data.dir = "~/Desktop/HT projects/Rosalind/Test"

# Task 1

# download file

ulr.test = "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz"

download.file(url = ulr.test, destfile = file.path(data.dir, "SRR020192.fastq.gz"))

# A Phred quality score is a measure of the quality of the identification of the nucleobases 
# generated by automated DNA sequencing.

# decompress file

# gunzip(file.path(data.dir, "SRR020192.fastq.gz"), remove = TRUE, overwrite = TRUE)

# read file

input <- file.path(data.dir, "SRR020192.fastq.gz")
file.name <- list.files(data.dir)
fq <- readFastq(input)

fq
# class: ShortReadQ
# length: 41892 reads; width: 40..280 cycles

fq <- fq[1:100]

head(sread(fq), 3)
# A DNAStringSet instance of length 3
# width seq
# [1]    74 GATGACGGTGTCTACATTGTTCCCGACCACTCATCTCCTCTGTCATGCCCGAAACGTCTTCTCAAACCCGTCGT
# [2]    66 GATGACGGTGTCTACATCGTTCCACCACTCATCTCCTCTGTCATGCCCGAAAGTCTTCTCAAACCC
# [3]   111 GACGACGGTGTCTACATCGTTCCACCACTCATCTCCTCTGTCATGCC...TTCGACAGTTTGGTCCTGAAACCGAACCCGGACCGACAAACGAACGT

head(quality(fq), 3)
# class: FastqQuality
# quality:
#   A BStringSet instance of length 3
# width seq
# [1]    74 98<???88988?=<::<<<779333?4488??AAA?988<?A?A>==2222922.02::998<<0000022,00
# [2]    66 <<<???889<<?=<<<<??<<<<?8888??AA??<<<<?>=>>>>200000002;0000<<00000
# [3]   111 AAAAAAAAAAAAAAAAAAIIIIIIIIIIIIIIIIIIIFFFAFFCC??...42799668<89770886699:000,,008998,,00000060,,...


myFilterAndTrim <- function(file, file.name, destination.directory)
{
  ## open input stream
  stream <- open(FastqStreamer(file))
  on.exit(close(stream))
  repeat {
    ## input chunk
    fq <- readFastq(input)
    fq <- fq[1:25]
    #      fq <- yield(stream)
    if (length(fq) == 0)
      break
    ## trim and filter, e.g., reads cannot contain 'N'...
    fq <- fq[nFilter()(fq)] # see ?srFilter for pre-defined filters
    ## trim as soon as 2 of 5 nucleotides has quality encoding less
    ## than "4" (phred score 20)
    fq <- trimTailw(fq, 2, "4", 2)
    ## drop reads that are less than 50nt
    fq <- fq[width(fq) >= 50]
    ## append to destination
    writeFastq(object = fq, 
               file = file.path(destination.directory, paste0("subset_", file.name)), 
               mode = "a")
  }
}

myFilterAndTrim(file = input, file.name = file.name, destination.directory = data.dir)


# Task 2


fq_filtered <- readLines(file.path(data.dir, "subset_SRR020192.fastq"))
fq_filtered <- read.table(file.path(data.dir, "subset_SRR020192.fastq"), fill = TRUE)

dir.create(data.dir, "SRR020192")
out.dir <- file.path(data.dir, "SRR020192")
setwd(out.dir)
getwd()

file.split(fn = "subset_SRR020192.fastq", size = 4, suf = "seq")
file.remove("subset_SRR020192.fastq")

list.reads = list.files(path = out.dir)

adapter = "CGTCCCA"

for (read in list.reads) {
  read <- readLines(read)
  name = read[1]
  seq = read[2]
  plus = read[3]
  quality = read[4]
  if (grepl(adapter, seq) == TRUE) {
    new.seq = sub(paste0(".*", adapter), "", seq)
    writeLines(text = c(name, new.seq, plus, quality), 
               con = file.path(out.dir, paste0("rm_adapter", new.seq,".fastq")),
               sep = "\n")
  } else {
    # writeLines(text = c(name, seq, plus, quality), 
    #            con = file.path(out.dir, paste0("rm_adapter", seq,".fastq")),
    #            sep = "\n")
    next()
  }
}


desired.reads = list.files(pattern = "m_adapter", path = out.dir, include.dirs = TRUE)

combined.desired.reads <- c()

for (read in desired.reads) {
  temp <- readLines(read)
  combined.desired.reads <- c(combined.desired.reads, temp)
}

combined.desired.reads





# processFile = function(filepath) {
#   con = file(filepath, "r")
#   while ( TRUE ) {
#     line = readLines(con, n = 1)
#     if ( length(line) == 0 ) {
#       break
#     }
#     print(line)
#   }
#   close(con)
# }
